### ISSUE TITLE
December Iteration Plan

### ISSUE THREAD

**Issue by egamma** [Microsoft/vscode#917]
After a major effort like the November Beta release the team does need some well deserved _decompression_. During this decompression period we still want to be responsive to issues raised by the community, but we also want to work on debt reduction, tune our own development process, and do some explorations. This also means that we will not start working on new big rock features, so work on code folding will not start yet, but do not worry, it will come.

For more information about our development process please refer to the development process [page](https://github.com/Microsoft/vscode/wiki/Development-Process).

Test Plan for this iteration: #1096

## Dates
The goal is to make the December update available before the holidays
- `stable` channel updated on Dec 18th
- `insider` channel updated on Dec 14th. This enables community members to test a candidate build at the same time as we do. For more information on the insider channel refer to [Test Prereleases](https://code.visualstudio.com/Docs/supporting/FAQ#_how-can-i-test-prerelease-versions-of-vs-code)

End game dates (we end an iteration with an [endgame](https://github.com/Microsoft/vscode/wiki/Development-Process))
- Friday Dec 11th evening: freeze
- Monday: test day, **update insider channel**
- Tuesday: fix day
- Wedneday: fix day
- Thursday: Done, documentation, release notes update, bugs verified
- Friday; **update the stable channel**

## Plan Items

**Performance**
We care a lot about performance...
- [x] Add telemetry to measure startup time #707 @SofianHn
  - We had a startup performance regression (#335) that required a [recovery build](https://github.com/Microsoft/vscode/releases/tag/0.10.3). We would like to avoid this in the future.
- [x] Investigate into a fix for the slow loading of files with multi byte characters #94 
  - [x] Investigate into a contribution for atom/node-oniguruma#45 @alexandrudima 
  - PR submitted atom/node-oniguruma#46

**Feedback**
Issues to address feedback from the community:
- Editor: It is difficult to deal with keybindings when using non-US standard keyboard layout #713 
 - [x] linux, windows @alexandrudima
 - [x] mac @alexandrudima @jrieken (offered to look into the Mac implementation)
- [x] API: Expose more editor commands as API #913 @jrieken
  - [x] fix language feature commands
  - [x]  documentation support and validation for parameter/return types
- [x] API: Investigate deactivate() on shutdown call for an extension @alexandrudima 
  - Issues: #537, #567, #941, #948, Microsoft/vscode-go#26
- [x] API: Support to contribute a JSON schema for a file pattern from an extension #489 @aeschli
  - This enables that extensions like jshint or eslint can contribute a schema for JSON validation for their configuration files.
  - It also enables to move more hard coded JSON schemas into the corresponding extensions
  - Issue: Microsoft/vscode-jshint#4 
- Deferred - API: Investigate into how we can provide more support for the VIM extensions @alexandrudima
  - Issues: #300, #575, #561, #1066
- [x] Gallery: Improve experience when user wants to read the README.md before installing an extension #914 @joaomoreno 
  - Issue: #727
- [x] Gallery: Improve performance of fetching the available extensions @joaomoreno
  - Issue: #1008
- [x] Git: Support git integration when git folder is in a parent folder #863 @joaomoreno
  - Issue: #863
- [x] Git: Add a pull with rebase action to the git viewlet #907 @joaomoreno
  - Issue: #150
- [x] JS: Simplify using ES6 @jrieken
  - Issue: #337
- [x] JS Enable ES6 coloring by using the TypeScript textmate grammar @aeschli
- [x] Debug: Improve presenting long text values inside the debug console @isidorn
  - Issues: #503

**Debt Items**
This section lists items to reduce our engineering debt:
- [x] Improve the internals of the suggestion contribution #1006 #1079 @jrieken @joaomoreno 
- Deferred - Proposal for separating the vscode tools from the vscode API of the existing vscode module #970 @bpasero
  - Currently the vscode node module combines the tools and the vscode API into one unit, we want to decouple the API from the used modules.
- Blocked - Automated Mac signing @joaomoreno 
  - The builds we publish need to be signed and this process isn't as automated yet, as we would like.
- [x] Increase test coverage @Microsoft/vscode 
- in progress - Debug: Provide a test suite for debugger adapters @weinand
- in progress - Language services - Document the language services protocol @dbaeumer

**Adoptions**
We continously have to adopt components which we require:
- [x] Adopt electron version 0.34.5 #826 @bpasero 
  - We usually update to new version of electron. This version addresses atom/electron#3349
- [x] Adopt [TypeScript 1.7](http://blogs.msdn.com/b/typescript/archive/2015/11/30/announcing-typescript-1-7.aspx) #836  @dbaeumer
  - Investigate into using async/await in extensions
- [x] Adopt the a newer version of OmniSharp @jrieken

**Engineering Items**
This section lists several engineering/development items that the team identified:
- [x] Run VSCode unit tests headless in a CI setup  @bpasero
  - Issues #912 #779  
- [x] Support a command line switch to launch Code without any extensions #708 @bpasero
  - Simplify diagnosing issues caused by extensions.
- [x] Move C# into a separate extension @jrieken
  - Enable the ASP/Roslyn team to eventually take over this extension
- [x] Extensions: warn the user when contributions conflict @isidorn 
- [x] Extension host debugging: change to a model where VS Code triggers reattach (instead of polling) #855 @weinand @bpasero @isidorn 
- [x] Support to define and preserve the zoom state of VS Code #291 @bpasero 
  - This was a real pain during the recent demos.
- [x] Git: Make it easy to submit a pull request/create and publish a local branch @joaomoreno
  - We do a lot of PRs and we want to simplify the flow.
- [x] Debug: Provide npm modules for the debugger protocol (client, server) @weinand
- [x] Complete the language service protocol and add more Code language features to the language service protocol @dbaeumer 
- Deferred - Debug: Provide function break point support in the UI and backend @isidorn @weinand
  - Conditional break points is a stretch goal for December.
- [x] What is the future of Send a smile @SofianHn 

**Explorations**
We also do some explorations:
- [x] Investigate into Salsa (Microsoft/TypeScript#4789) adoption with the TypeScript team
- In progress - Explore integrated TypeScript building/transpilation #973 @dbaeumer
- Under discussion - Explore: Improved Intellisense support to show documentation for a  completion item #369  @joaomoreno 
- [x] Explore support for fuzzy searching in the UI #968

## Bugs
This section includes bug queries for the vscode related repositories:
- [vscode](https://github.com/Microsoft/vscode/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Abug+milestone%3A%22Dec+2015%22+-label%3Aupstream+)
- [vscode-vsce](https://github.com/Microsoft/vscode-vsce/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Abug+milestone%3A%22Dec+2015%22+-label%3Aupstream+)
- [vscode-node-debug](https://github.com/Microsoft/vscode-node-debug/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Abug+milestone%3A%22Dec+2015%22+-label%3Aupstream+)
- [vscode-mono-debug ](https://github.com/Microsoft/vscode-mono-debug/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Abug+milestone%3A%22Dec+2015%22+-label%3Aupstream+)
- [vscode-loader](https://github.com/Microsoft/vscode-loader/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Abug+milestone%3A%22Dec+2015%22+-label%3Aupstream+)
- [vscode-textmate](https://github.com/Microsoft/vscode-textmate/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Abug+milestone%3A%22Dec+2015%22+-label%3Aupstream+)
- [vscode-generator-code](https://github.com/Microsoft/vscode-generator-code/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Abug+milestone%3A%22Dec+2015%22+-label%3Aupstream+)
- [vscode-languageserver-node](https://github.com/Microsoft/vscode-generator-code/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3Abug+milestone%3A%22Dec+2015%22+-label%3Aupstream+)

**Comment by Stanzilla** [comment]
Sneak in https://github.com/Microsoft/vscode/issues/507 maybe?

**Comment by egamma** [comment]
@Stanzilla thanks for reviewing, request makes sense, but let's do the _sneaking_ once all the other work is done.
